src/core/control/xojfile/SaveHandler.cpp

```diff
@@ -269,33 +269,46 @@ void SaveHandler::visitPage(XmlNode* root, PageRef p, Document* doc, int id) {
             char* filename = g_strdup_printf("bg_%d.png", this->attachBgId++);
             background->setAttrib("domain", "attach");
             background->setAttrib("filename", filename);
-            p->getBackgroundImage().setFilepath(filename);
 
             backgroundImages.emplace_back(p->getBackgroundImage());
 
+            /*
+             * Because BackgroundImage is basically a wrapped pointer, the following lines actually modify
+             * *(p->getBackgroundImage().content) and thus the Document.
+             * TODO Find a better way
+             */
+            backgroundImages.back().setFilepath(filename);
+            backgroundImages.back().setCloneId(id);
```

Comment:
I fear to lock the whole document at once for a single change. This will make the hole application single threaded and will also block most of the UI at each operation. Instead, I would like to have many independent subgroups of the document, which can be locked and modified simultaneously.

Another alternative would be, to remove the locking completely and to replace any changeable subpart by an atomic<shared_ptr> to constant (immutable) Data. The data can then be replaced with a read-modify-write operation.

Reply:
> > Another alternative would be, to remove the locking completely and to replace any changeable subpart by an atomic<shared_ptr> to constant (immutable) Data. The data can then be replaced with a read-modify-write operation.
> 
> This would be very expensive. I don't quite see how to clone the content of a layer in an atomic way either.

It hardly depends on the data structure. Every change must be kept in the Undo Redo Handler, therefore it is copied anyway. When you now store all the document state in a tree of shared_ptr nodes, you would just reference the unit you want to change in the undo-redo handler and then replace the shared_ptr with a copy of it and its data. The replacement of the shared_ptr can be done atomically. We only need a solution for the ABA problem, but that's solvable by introducing a change index.